lib/Dialect/Arc/Transforms/MergeIfs.cpp

```diff
@@ -0,0 +1,365 @@
+//===- MergeIfs.cpp -------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Arc/ArcOps.h"
+#include "circt/Dialect/Arc/ArcPasses.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "arc-merge-ifs"
+
+namespace circt {
+namespace arc {
+#define GEN_PASS_DEF_MERGEIFSPASS
+#include "circt/Dialect/Arc/ArcPasses.h.inc"
+} // namespace arc
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+using namespace arc;
+
+namespace {
+struct MergeIfsPass : public arc::impl::MergeIfsPassBase<MergeIfsPass> {
+  void runOnOperation() override;
+  using MergeIfsPassBase::MergeIfsPassBase;
+  using MergeIfsPassBase::numIfsMerged;
+  using MergeIfsPassBase::numOpsMovedFromBetweenIfs;
+  using MergeIfsPassBase::numOpsMovedToUser;
+  using MergeIfsPassBase::numOpsSunk;
+};
+
+/// A helper to perform the op sinking within a specific block.
+struct Sinker {
+  MergeIfsPass &pass;
+  Block &rootBlock;
+  bool anyChanges;
+
+  Sinker(MergeIfsPass &pass, Block &rootBlock)
+      : pass(pass), rootBlock(rootBlock) {}
+  LogicalResult run();
+  void sinkOps();
+  void mergeIfs();
+};
+} // namespace
+
+void MergeIfsPass::runOnOperation() {
+  // Go through the regions recursively, from outer regions to nested regions,
+  // and try to move/sink/merge ops in each.
+  auto result = getOperation()->walk<WalkOrder::PreOrder>([&](Region *region) {
+    if (region->hasOneBlock() && mlir::mayHaveSSADominance(*region))
+      if (failed(Sinker(*this, region->front()).run()))
+        return WalkResult::interrupt();
+    return WalkResult::advance();
+  });
+  if (result.wasInterrupted())
+    signalPassFailure();
+}
+
+/// Iteratively sink ops into block and move them closer to their uses, and
+/// merge adjacent `scf.if` operations.
+LogicalResult Sinker::run() {
+  LLVM_DEBUG(llvm::dbgs() << "Running on block in "
+                          << rootBlock.getParentOp()->getName() << "\n");
+  do {
+    anyChanges = false;
+    sinkOps();
+    mergeIfs();
+  } while (anyChanges);
```

Comment:
Nothing important, but the number of iterations here might make for another nice pass statistic 🙂

Reply:
Haha I had that but removed it again 😅. Let me add it back in -- it's good to have around!